controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java

```diff
@@ -169,17 +175,19 @@ protected void startUp() {
 
             AuthHelper authHelper = new AuthHelper(serviceConfig.getGRPCServerConfig().get().isAuthorizationEnabled(),
                     serviceConfig.getGRPCServerConfig().get().getTokenSigningKey());
-            streamMetadataTasks = new StreamMetadataTasks(streamStore, hostStore, taskMetadataStore,
+            streamMetadataTasks = new StreamMetadataTasks(streamStore, bucketStore, hostStore, taskMetadataStore,
                     segmentHelper, controllerExecutor, host.getHostId(), connectionFactory, authHelper, requestTracker);
             streamTransactionMetadataTasks = new StreamTransactionMetadataTasks(streamStore,
                     hostStore, segmentHelper, controllerExecutor, host.getHostId(), serviceConfig.getTimeoutServiceConfig(),
                     connectionFactory, authHelper);
+            
+            BucketServiceFactory bucketServiceFactory = new BucketServiceFactory(host.getHostId(), bucketStore, streamStore, 
+                    streamMetadataTasks, retentionExecutor, requestTracker);
+            retentionService = bucketServiceFactory.getBucketManagerService(BucketStore.ServiceType.RetentionService);
```

Comment:
A very large amount of the code in this PR is working around this circular dependency.
This is an anti-pattern and this code review is a good example of why. Because the RetentionService depends on streamMetadataTasks which depend on the BucketStore which in turn depends on the retention service (through the listener interface) all of these needed to be made into services, listeners need to be created and registered and unregistered, and there is a queue and a thread pool. 

If you refactor the flow, this should simplify a lot. There can be a task the scans the store and is registered for callbacks and it can invoke the correct manager. It can be instantiated last. Then there is only one service/threadpool and no queuing or registration.

Reply:
Let me see if i understand this correctly --> What you are suggesting is that we have 4 objects instead of 3. 
1. zkStore
2. streamMetadataTasks --> uses zkstore
3. manager service --> uses streamMetadataTasks, zk store
4. some new task --> uses zkstore, retention manager

retention manager registers callbacks on the new `task`. 
This task periodically scans the store and calls into manager service. 

Right?

There is one problem though.. we have two levels of listeners -
1. bucket ownership
2. streams under buckets

Using the suggested approach `2` can be taken care of, but for `1` we will still need zkstore to callback into either task or manager services. 

Let me think about it some more. 
Open to more ideas. @tkaitchuck

Reply:
The problem here is that the BucketStore really has two unrelated responsibilities, leader election, and notifying of new streams. Let's separate those. We could define an interface BucketOwnershipManager which does leader election and has a ZK backed impl, and a in memory stub. Then there should be a separate class that provides an abstraction over the actual stored data. We can call that BucketStore. There can be a ZK impl and an in memory impl. Either way they should provide the ability to fetch appropriate info from the actual store.

Once we do that the problem breaks down as follows:
There is a retention manager that exposes an appropriate interface. 
The store exposes some sort of iterator/read abstraction.
The leader elector can invoke callbacks that are provided to it at its construction.

Then we construct a `Dispatcher` object that has a callback take ownership of a bucket. It has a reference to the store. During the callback it invokes operations on the store, gets the data and invokes the retention manager (which it owns).

The dispatcher is passed to the leader elector in it's constructor. 

So the references go:
LeaderElector -> Dispatcher
Dispatcher -> Store
Dispatcher -> Retention Manger.

And either the Dispatcher or the Retention manager have a thread pool, but no need for both.

Reply:
I have made the following changes now:
BucketManager and BucketService are abstract classes and can be instantiated in two flavours - inMemory and Zookeeper implementations. 
The factory does that based on the type of bucket store used. 
This means there is no cyclic dependency per se. 

Bucket store now only contains store related apis and not the listeners. 

RetentionWorkflow is supplied as a functional interface to the bucketManger (to the factory) and same infrastructure could be used for multiple types of workflows by just supplying the correct workflow method at the time of factory instantiation.